docs/evm/guides/vrf.md

```diff
@@ -0,0 +1,288 @@
+
+    struct Bet {
+        uint256 amount;
+        bytes32 commitment;
+        bool revealed;
+        uint64 result;
+    }
+
+    mapping(address => Bet) public bets;
+
+    // Step 1: Commit the bet by submitting a hash (commit phase)
+    function commitBet(bytes32 commitment) public payable {
+        require(msg.value > 0, "Bet amount must be greater than zero");
+        require(bets[msg.sender].amount == 0, "Already committed a bet");
+
+        bets[msg.sender] = Bet({
+            amount: msg.value,
+            commitment: commitment,
+            revealed: false,
+            result: 0
+        });
+    }
+
+    // Step 2: Reveal the bet by providing the preimage (reveal phase)
+    function revealBet(uint64 nonce, uint64 guess) public {
+        Bet storage bet = bets[msg.sender];
+        require(bet.amount > 0, "No bet to reveal");
+        require(!bet.revealed, "Bet already revealed");
+
+        // Check that the commitment matches the hash
+        require(keccak256(abi.encodePacked(nonce, guess)) == bet.commitment, "Invalid commitment");
+
+        // Fetch a secure random number from Cadence Arch
+        (bool ok, bytes memory data) = cadenceArch.staticcall(abi.encodeWithSignature("revertibleRandom()"));
+        require(ok, "Failed to fetch a random number");
+        uint64 randomNumber = abi.decode(data, (uint64));
+
+        // Determine the result (0 = loss, 1 = win)
+        bet.result = randomNumber % 2;
+        bet.revealed = true;
+
+        // Pay the winner double the bet amount if they won
+        if (bet.result == guess) {
+            payable(msg.sender).transfer(bet.amount * 2);
+        }
+    }
+
+    // To reset and clean up after the reveal
+    function resetBet() public {
+        require(bets[msg.sender].revealed, "Bet not yet revealed");
+        delete bets[msg.sender];
+    }
+
+    // Helper function to generate the commitment hash (off-chain or in Remix)
+    function getHash(uint64 nonce, uint64 guess) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(nonce, guess));
+    }
+}
+
+```
+
+---
+
+### **Steps to Interact With the Coin Toss Contract**
+
+**Step 1: Commit Phase**
+
+You need to generate a **commitment hash** from a nonce (random number) and your guess (0 for heads, 1 for tails).
+
+1. Call the `getHash` function to generate the commitment hash:
+
+   ```solidity
+   getHash(12345, 1);  // nonce: 12345, guess: 1 (tails)
+   ```
+
+   Example Output: `0x5f5766a7a5d18bb9a9c3040a7b672c5c8f4860018ebd5a11f1678f11f2326c44`.
+
+2. Call **`commitBet(bytes32 commitment)`**:
+   - Input the generated commitment hash into the **commitBet** function.
+   - Enter the amount of **FLOW** to bet in the **Value input box** in Remix.
+   - Example: Enter `1` to bet **1 FLOW**.
+
+**Step 2: Reveal Phase**
+
+1. Call **`revealBet(uint64 nonce, uint64 guess)`** to reveal your bet:
+
+   ```solidity
+   revealBet(12345, 1);  // nonce: 12345, guess: 1 (tails)
```

Comment:
⚠️ If I understood this correctly, a user can call `revealBet` and decide to revert the result. The user can run `revealBet` in a later block using a different randomness, and try till they win. This is the problem described in `revertibleRandom`, and this scheme uses `revertibleRandom` too.

Reply:
That's a proper reading @tarakby. I overlooked that this contract uses `revertibleRandom` and assumed that it was using `getRandomSource` in my review